### PR TITLE
Landan/text encoder refactor

### PR DESCRIPTION
**Pull request by Landanjs**
This PR has a few new features:
1. Changes `SDXLTextEncoder` and `SDXLTokenizer` to be `MultiTextEncoder` and `MultiTokenizer`, respectively. These two new classes enable using arbitrary combinations of OpenAI CLIP, OpenCLIP, E5 and possibly other models with minor tweaks.
    a. **CONFIG CHANGE**: Since there is no more SDXL specific tokenizer, a new argument `sdxl_conditioning` is added in `image_caption.py` to specify whether to generate SDXL-style conditioning in the dataset.
2. Use the `attention_mask` in the text encoder. 

Minor:
1. Set `qkv_clip = None`

### CODE REVIEW

File: diffusion/datasets/image_caption.py [hunk 1]
```diff
@@ -154,31 +163,26 @@ def __getitem__(self, index):
...
+        for tokenizer in self.tokenizers:
+            tokenizer_out = tokenizer(caption,
+                                      padding='max_length',
+                                      max_length=tokenizer.model_max_length,
+                                      truncation=True,
+                                      return_tensors='pt')
+
+            out['captions'].append(tokenizer_out.input_ids.squeeze())
+            out['attention_mask'].append(tokenizer_out.attention_mask)
+        # TODO: do I need to take the union of the masks?
```
Comment (jazcollins) [inline diffusion/datasets/image_caption.py:177]: yes i think so, because this gets passed into the unet as `encoder_attention_mask` which is expected to be (batch, sequence_length)

File: diffusion/models/models.py [hunk 2]
```diff
@@ -272,9 +273,38 @@ def stable_diffusion_xl(
...
+
+    torch_dtype = torch.float16 if encode_latents_in_fp16 else None
+    text_encoders = torch.nn.ModuleList()
+    for text_encoder_name in text_encoder_names:
+        # If tokenizer_name_or_path contains more than one '/', then the string includes a subfolder to extract
+        path_split = text_encoder_name.split('/')
+        name = '/'.join(path_split[:2])
+        subfolder = '/'.join(path_split[2:]) if len(path_split) > 2 else ''
+
+        # I'm not sure how to load this CLIP Text model using Auto...
```
Comment (jazcollins) [inline diffusion/models/models.py:302]: one way to achieve this (not sure if it is the cleanest way, so take it or leave it):

```
    torch_dtype = torch.float16 if encode_latents_in_fp16 else None
    text_encoders = torch.nn.ModuleList()
    d_text_encoder = 0
    d_model_keys = ['projection_dim', 'd_model', 'hidden_size'] # CLIP, T5, E5
    for text_encoder_name in text_encoder_names:
        # If tokenizer_name_or_path contains more than one '/', then the string includes a subfolder to extract
        path_split = text_encoder_name.split('/')
        name = '/'.join(path_split[:2])
        subfolder = '/'.join(path_split[2:]) if len(path_split) > 2 else ''

        # Get architecture and d_model for text encoder(s)
        text_encoder_config = PretrainedConfig.get_config_dict(name, subfolder=subfolder)[0] 
        
        d_model = None
        for d_model_key in d_model_keys:
            if d_model_key in text_encoder_config:
                # Add current text encoder d_model to total d_text_encoder
                d_text_encoder += text_encoder_config[d_model_key]
                break
        if not d_model:
            raise ValueError('Could not find any of d_model_keys in text encoder config')

        architectures = text_encoder_config['architectures']
        if architectures == ['CLIPTextModel']:
            text_encoders.append(CLIPTextModel.from_pretrained(name, subfolder=subfolder, torch_dtype=torch_dtype))
        elif architectures == ['CLIPTextModelWithProjection']:
            text_encoders.append(CLIPTextModelWithProjection.from_pretrained(name, subfolder=subfolder, torch_dtype=torch_dtype))
        else:
            text_encoders.append(AutoModel.from_pretrained(name, subfolder=subfolder, torch_dtype=torch_dtype))
```

this also loads each text encoder's config and tries to find the key that corresponds to the model output dim - will need this later to set the `cross_attention_dim` for the unet

**Review by jazcollins — COMMENTED** [review]
left some comments!!

**Review by jazcollins — COMMENTED** [review]
thanks for refactoring!! 

since we newly have the MultiTokenizer, we will want to change this line as well to access one of the tokenizers in the list:
https://github.com/mosaicml/diffusion/blob/fa13a3056bf0baa0612416cf67a39990178a3827/diffusion/evaluation/clean_fid_eval.py#L163

File: diffusion/datasets/image_caption.py [hunk 3]
```diff
@@ -42,26 +42,27 @@ class StreamingImageCaptionDataset(StreamingDataset):
         transform (Callable, optional): The transforms to apply to the image. Default: ``None``.
         image_key (str): Key associated with the image in the streaming dataset. Default: ``'image'``.
         caption_key (str): Key associated with the caption in the streaming dataset. Default: ``'caption'``.
-        sdxl (bool): Whether or not we're training SDXL. Default: `False`.
+        sdxl_conditioning (bool): Whether or not to include SDXL microconditioning in a sample. Default: `False`.
         zero_dropped_captions (bool): If True, zero out text embeddings for dropped captions. Default: ``False``.
-
         **streaming_kwargs: Additional arguments to pass in the construction of the StreamingDataloader
+
+        TODO: Add a return documentation
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply diffusion/datasets/image_caption.py]: Oops forgot to push!

File: diffusion/models/text_encoder.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            name_split = model_name.split('/')
+            base_name = '/'.join(name_split[:2])
+            subfolder = '/'.join(name_split[2:])
+            text_encoder_config = PretrainedConfig.get_config_dict(base_name, subfolder=subfolder)[0]
+
+            # Add text_encoder output dim to total dim
+            dim_found = False
+            for model_dim_key in model_dim_keys:
+                if model_dim_key in text_encoder_config:
+                    self.text_encoder_dim += text_encoder_config[model_dim_key]
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply diffusion/models/text_encoder.py:58]: Ah, thanks for catching this! I'll delete `projection_dim` for now since I think that could be independent of the text embedding size if that make sense to you as well?

File: diffusion/models/text_encoder.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            else:
+                self.text_encoders.append(
+                    AutoModel.from_pretrained(base_name, subfolder=subfolder, torch_dtype=torch_dtype))
+            self.architectures += architectures
+
+    @property
+    def device(self):
+        return self.text_encoders[0].device
+
+    def forward(self, tokenized_texts: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply diffusion/models/text_encoder.py]: Done! I had to add quiet a bit of masking code, but hopefully it is right now. It would be helpful if you can check my commits since your comment which will contain the masking code

File: diffusion/models/stable_diffusion.py [hunk 6]
```diff
@@ -142,52 +143,52 @@ def set_rng_generator(self, rng_generator: torch.Generator):
...
                 else:
-                    conditioning = self.text_encoder(conditioning)[0]
+                    latents = self.vae.encode(inputs)['latent_dist'].sample().data
+                # Encode tokenized prompt into embedded text and pooled text embeddings
+                text_encoder_out = self.text_encoder(conditionings, attention_mask=attention_mask)
+                text_embeds = text_encoder_out[0]
+                if self.sdxl:
+                    if len(text_encoder_out) <= 1:
+                        raise RuntimeError('SDXL requires text encoder output to include a pooled text embedding')
+                    text_pooled_embeds = text_encoder_out[1]
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:182]: The safest option would be to specify a key to index for the text embeddings and the projection embedding, but this would add two more arguments to the model and this PR is already pretty long...

File: diffusion/models/stable_diffusion.py [hunk 7]
```diff
@@ -448,45 +448,44 @@ def _prepare_text_embeddings(self, prompt, tokenized_prompts, tokenized_pad_mask
...
-                    if self.mask_pad_tokens:
-                        # For cross attention mask, take union of masks (want [B, 77])
-                        tokenized_pad_mask = torch.logical_or(tokenized_pad_mask[0], tokenized_pad_mask[1]).to(
-                            tokenized_pad_mask[0].dtype).to(device)
+                tokenized_prompts = tokenized_out['input_ids']
+                tokenized_pad_mask = tokenized_out['attention_mask']
+            if tokenized_pad_mask is not None:
+                tokenized_pad_mask = tokenized_pad_mask.to(device)
+            text_encoder_out = self.text_encoder(tokenized_prompts.to(device), attention_mask=tokenized_pad_mask)
+            prompt_embeds = text_encoder_out[0]
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:462]: Again, would be better if this was a key instead of index...

**Comment by Landanjs** [comment]
Ooops, requested review before implementing `.batch_decode`...

File: diffusion/models/stable_diffusion.py [hunk 8]
```diff
@@ -142,52 +143,52 @@ def set_rng_generator(self, rng_generator: torch.Generator):
...
+        latents, text_embeds, text_pooled_embeds, attention_mask, encoder_attention_mask = None, None, None, None, None
+        if 'attention_mask' in batch:
+            attention_mask = batch['attention_mask']  # mask for text encoders
+            # text mask for U-Net
+            if self.mask_pad_tokens:
+                if len(attention_mask.shape) == 2:
+                    encoder_attention_mask = attention_mask
+                elif len(attention_mask.shape) == 3:
+                    encoder_attention_mask = attention_mask[:, 0]
+                    for i in range(1, attention_mask.shape[1]):
```
Comment (coryMosaicML) [inline diffusion/models/stable_diffusion.py:155]: This should probably be done in dataloading. No need to change it for this PR though.

**Review by coryMosaicML — APPROVED** [review]
Looks good! Thanks for doing this!

File: diffusion/models/stable_diffusion.py [hunk 9]
```diff
@@ -142,52 +143,52 @@ def set_rng_generator(self, rng_generator: torch.Generator):
...
-                    else:
-                        conditioning = self.text_encoder(conditioning)[0]  # Should be (batch_size, 77, 768)
+            inputs, conditionings = batch[self.image_key], batch[self.text_key]
 
-            else:
-                latents = self.vae.encode(inputs)['latent_dist'].sample().data
-                if self.sdxl:
-                    conditioning, pooled_conditioning = self.text_encoder([conditioning, conditioning_2])
+            # If encode_latents_in_fp16, disable autocast context as models are in fp16
+            c = torch.cuda.amp.autocast(enabled=False) if self.encode_latents_in_fp16 else nullcontext()  # type: ignore
```
Comment (Skylion007) [inline diffusion/models/stable_diffusion.py:169]: i'm a bit confused should it be `enabled=not self.encode_latents_in_fp16` or something?

File: diffusion/models/stable_diffusion.py [hunk 10]
```diff
@@ -448,45 +448,44 @@ def _prepare_text_embeddings(self, prompt, tokenized_prompts, tokenized_pad_mask
...
+                tokenized_pad_mask = tokenized_pad_mask.to(device)
+            text_encoder_out = self.text_encoder(tokenized_prompts.to(device), attention_mask=tokenized_pad_mask)
+            prompt_embeds = text_encoder_out[0]
             if self.sdxl:
-                text_embeddings, pooled_text_embeddings = self.text_encoder(
-                    [tokenized_prompts[:, 0, :].to(device), tokenized_prompts[:, 1, :].to(device)])  # type: ignore
-            else:
-                text_embeddings = self.text_encoder(tokenized_prompts.to(device))[0]  # type: ignore
+                if len(text_encoder_out) <= 1:
+                    raise RuntimeError('SDXL requires text encoder output to include a pooled text embedding')
```
Comment (Skylion007) [inline diffusion/models/stable_diffusion.py:465]:
```suggestion
                    raise ValueError('SDXL requires text encoder output to include a pooled text embedding')
```

File: diffusion/models/text_encoder.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+    def forward(self, input_ids: torch.Tensor, attention_mask: Optional[torch.Tensor] = None):
+        # Check input_ids and attention_mask is shape [batch_size, len(self.text_encoders), max_sequence_length]
+        if len(input_ids.shape) == 2:
+            input_ids = input_ids.unsqueeze(dim=1)
+        if attention_mask is not None and len(attention_mask.shape) == 2:
+            attention_mask = attention_mask.unsqueeze(dim=1)
+        if input_ids.shape[1] != len(self.text_encoders) or (attention_mask is not None and
+                                                             attention_mask.shape[1] != len(self.text_encoders)):
+            raise RuntimeError(
```
Comment (Skylion007) [inline diffusion/models/text_encoder.py:95]:
```suggestion
            raise ValueError(
```

File: diffusion/models/stable_diffusion.py (hunk 10 above)
Comment (Landanjs) [reply diffusion/models/stable_diffusion.py:465]: Should this be a ValueError? The error is from using a text encoder that doesn't output a pooled embedding

File: diffusion/models/stable_diffusion.py (hunk 9 above)
Comment (Landanjs) [reply diffusion/models/stable_diffusion.py:169]: TBH, I'm not sure, Mihir added this code, so I was trying to keep the logic the same while condensing it a bit

File: diffusion/models/stable_diffusion.py (hunk 8 above)
Comment (Landanjs) [reply diffusion/models/stable_diffusion.py:155]: Yeah, that makes sense 😓

**Review by jazcollins — APPROVED** [review]
lgtm, this is awesome landan!!! 

doesn't necessarily have to be in the same PR, but it would be good to update the SDXL sample yamls (both [mosaic](https://github.com/mosaicml/diffusion/tree/main/yamls/mosaic-yamls) and [hydra](https://github.com/mosaicml/diffusion/tree/main/yamls/hydra-yamls)) with the new relevant yaml params (`tokenizer_names_or_paths` and `sdxl_conditioning` in the dataloader, etc.)
